source/fuzz/fuzzer_pass_duplicate_regions_with_selections.cpp

```diff
@@ -0,0 +1,128 @@
+    // Randomly decide whether to apply the transformation.
+    if (!GetFuzzerContext()->ChoosePercentage(
+            GetFuzzerContext()->GetChanceOfDuplicatingRegionWithSelection())) {
+      continue;
+    }
+    std::vector<opt::BasicBlock*> start_blocks;
+    for (auto& block : *function) {
+      // Currently, we don't consider the first block to be the starting block.
+      if (&block == &*function->begin()) {
+        continue;
+      }
+      start_blocks.push_back(&block);
+    }
+    if (start_blocks.empty()) {
+      continue;
+    }
+    // Randomly choose the entry block.
+    auto entry_block =
+        start_blocks[GetFuzzerContext()->RandomIndex(start_blocks)];
+    auto dominator_analysis = GetIRContext()->GetDominatorAnalysis(function);
+    auto postdominator_analysis =
+        GetIRContext()->GetPostDominatorAnalysis(function);
+    std::vector<opt::BasicBlock*> candidate_exit_blocks;
+    for (auto postdominates_entry_block = entry_block;
+         postdominates_entry_block != nullptr;
+         postdominates_entry_block = postdominator_analysis->ImmediateDominator(
+             postdominates_entry_block)) {
+      // The candidate exit block must be dominated by the entry block and the
+      // entry block must be post-dominated by the candidate exit block. Ignore
+      // the block if it heads a selection construct or a loop construct.
+      if (dominator_analysis->Dominates(entry_block,
+                                        postdominates_entry_block) &&
+          !postdominates_entry_block->GetLoopMergeInst()) {
+        candidate_exit_blocks.push_back(postdominates_entry_block);
+      }
+    }
+    if (candidate_exit_blocks.empty()) {
+      continue;
+    }
+    // Randomly choose the exit block.
+    auto exit_block = candidate_exit_blocks[GetFuzzerContext()->RandomIndex(
+        candidate_exit_blocks)];
+
+    auto region_blocks =
+        TransformationDuplicateRegionWithSelection::GetRegionBlocks(
+            GetIRContext(), entry_block, exit_block);
+
+    // Construct |original_label_to_duplicate_label| by iterating over all
+    // blocks in the region. Construct |original_id_to_duplicate_id| and
+    // |original_id_to_phi_id| by iterating over all instructions in each block.
+    std::map<uint32_t, uint32_t> original_label_to_duplicate_label;
+    std::map<uint32_t, uint32_t> original_id_to_duplicate_id;
+    std::map<uint32_t, uint32_t> original_id_to_phi_id;
+    for (auto& block : region_blocks) {
+      original_label_to_duplicate_label[block->id()] =
+          GetFuzzerContext()->GetFreshId();
+      for (auto& instr : *block) {
+        if (instr.result_id()) {
+          original_id_to_duplicate_id[instr.result_id()] =
+              GetFuzzerContext()->GetFreshId();
+          if ((&instr == &*exit_block->tail() ||
+               fuzzerutil::IdIsAvailableBeforeInstruction(
+                   GetIRContext(), &*exit_block->tail(), instr.result_id()))) {
```

Comment:
Please insert a brief comment before this conditional explaining what's going on - i.e., that exit_block's tail is the final instruction in the region, so it or anything available right before it will be available after the region.

Also I think &*exit_block->tail is a nasty-enough expression that it merits having a local variable devoted to it right before the conditional.